irods/at_client_exit.py

```diff
@@ -0,0 +1,110 @@
+                    if notify:
+                        notify(function, stage)
+                    function()
+                except Exception as exc:
+                    logger.warning("%r raised from %s", exc, function)
+    finally:
+        _reset_cleanup_functions()
+
+# The primary interface: 'before' and 'after' are the execution order relative to the DURING stage,
+# in which the Python client cleans up its own session objects and runs the optional data object auto-close facility.
+
+def register_for_execution_before_prc_cleanup(function, **kw_args):
+    """Register a function to be called by the Python iRODS Client (PRC) at exit time.
+       It will be called without arguments, and before PRC begins its own cleanup tasks."""
+    return _register(LibraryCleanupStage.BEFORE, function, **kw_args)
+
+def register_for_execution_after_prc_cleanup(function, **kw_args):
+    """Register a function to be called by the Python iRODS Client (PRC) at exit time.
+       It will be called without arguments, and after PRC is done with its own cleanup tasks."""
+    return _register(LibraryCleanupStage.AFTER, function, **kw_args)
+
+class unique_function_invocation:
+    """Wrap a function object to provide a unique handle for registration of a given
+       function multiple times, possibly with different arguments.
+    """
+    def __init__(self, func, tup_args = (), kw_args = ()):
+        self.tup_args = tup_args
+        self.kw_args = dict(kw_args)
+        self.func = func
+        self.at_exit_stage = None
+
+    def __call__(self):
+        return (self.func)(*self.tup_args, **self.kw_args)
+
+def get_stage(n = 2):
+    """A utility function that can be called from within the highest call frame (use a higher
+       "n" for subordinate frames) of a user-defined cleanup function to get the currently active
+       cleanup stage, defined by a LibraryCleanupStage enum value.  The user-defined function must have been
+       wrapped by an instance of class unique_function_invocation, and that wrapper object
+       registered using the keyword argument: stage_notify_function = NOTIFY_VIA_ATTRIBUTE."""
+    caller = sys._getframe(n).f_locals.get('self')
+    if isinstance(caller, unique_function_invocation):
+        return caller.at_exit_stage
+
+if __name__ == '__main__':
+
+    def myfunction(message='the_default_message'):
+        print(f"\n{message = }")
+        print(f"{get_stage() = }")
+
+    register_for_execution_before_prc_cleanup(unique_function_invocation(myfunction, ['#1 pre']), stage_notify_function = NOTIFY_VIA_ATTRIBUTE)
+    register_for_execution_before_prc_cleanup(unique_function_invocation(myfunction, ['#2 pre']))
+
+    register_for_execution_after_prc_cleanup(unique_function_invocation(myfunction, ['post']), stage_notify_function = NOTIFY_VIA_ATTRIBUTE)
+    register_for_execution_after_prc_cleanup(myfunction)
+
+    print('-- prepare to see more printed out during cleanup! --')
```

Comment:
What is this for? Testing?

Reply:
Technically it should be a test, yes.

Reply:
Will convert it momentarily. It's been manually run and passed. It's a verification that a function object can get notification of the stage in which it's being run.

Reply:
Sounds good.